src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs

```diff
@@ -1195,17 +1465,739 @@ private static void ReportReadOnlyError(Symbol symbol, SyntaxNode node, BindValu
+                    {
+                        break;
+                    }
+
+                    return CheckInvocationEscape(
+                        indexerAccess.Syntax,
+                        indexerSymbol,
+                        indexerAccess.ReceiverOpt,
+                        indexerSymbol.Parameters,
+                        indexerAccess.Arguments,
+                        indexerAccess.ArgumentRefKindsOpt,
+                        indexerAccess.ArgsToParamsOpt,
+                        checkingReceiver,
+                        escapeFrom,
+                        escapeTo,
+                        diagnostics);
+
+                case BoundKind.PropertyAccess:
+                    var propertyAccess = (BoundPropertyAccess)expr;
+                    var propertySymbol = propertyAccess.PropertySymbol;
+
+                    if (propertySymbol.RefKind == RefKind.None)
+                    {
+                        break;
+                    }
+
+                    // not passing any arguments/parameters
+                    return CheckInvocationEscape(
+                        propertyAccess.Syntax,
+                        propertySymbol,
+                        propertyAccess.ReceiverOpt,
+                        default,
+                        default,
+                        default,
+                        default,
+                        checkingReceiver,
+                        escapeFrom,
+                        escapeTo,
+                        diagnostics);
+            }
+
+            // At this point we should have covered all the possible cases for anything that is not a strict RValue.
+            Error(diagnostics, GetStandardRValueRefEscapeError(escapeTo), node);
+            return false;
+        }
+
+        /// <summary>
+        /// Computes the widest scope depth to which the given expression can escape by value.
+        /// 
+        /// NOTE: unless the type of expression is ref-like, the result is Binder.ExternalScope since ordinary values can always be returned from methods. 
+        /// </summary>
+        internal static uint GetValEscape(BoundExpression expr, uint scopeOfTheContainingExpression)
+        {
+            // cannot infer anything from errors
+            if (expr.HasAnyErrors)
+            {
+                return Binder.ExternalScope;
+            }
+
+            // constants/literals cannot refer to local state
+            if (expr.ConstantValue != null)
+            {
+                return Binder.ExternalScope;
+            }
+
+            // to have local-referring values an expression must have a ref-like type
+            if (expr.Type?.IsByRefLikeType != true)
+            {
+                return Binder.ExternalScope;
+            }
+
+            // cover case that can refer to local state
+            // otherwise default to ExternalScope (ordinary values)
+            switch (expr.Kind)
+            {
+                case BoundKind.DefaultExpression:
```

Comment:
Should those cases be removed, since we will fall back to ExternalScope below anyways? 
Alternatively, should this switch be made comprehensive instead (default: throw UnexpectedException();)?

Reply:
I wanted to call out these, since it may be not obvious. Otherwise, yes, they could be removed. 
We should consider a comprehensive switch eventually. We are not handling all cases yet though.

---
In reply to: [135364103](https://github.com/dotnet/roslyn/pull/21334#discussion_r135364103) [](ancestors = 135364103)